### ISSUE TITLE
Version 3.8.0 comma-dangle breaks Flowtype annotations when a object rest parameter is used

### ISSUE THREAD

**Issue by mormahr** [eslint/eslint#7370]
**Tell us about your environment**

* **ESLint Version:**
v3.8.0
* **Node Version:**
v6.7.0
* **npm Version:**
3.10.3

**What parser (default, Babel-ESLint, etc.) are you using?**
Babel-ESLint

**Please show your full configuration:**

.eslintrc.js
```js
module.exports = {
	"parser": "babel-eslint",
	"parserOptions": {
		"ecmaVersion": 6,
		"sourceType": "module"
	},
	"env": {
		"es6": true,
		"node": true
	},
	"plugins": [
		"flowtype",
		"flow-vars"
	],
	"extends": "eslint:recommended",
	"rules": {
		"comma-dangle": [
			"error",
			"always-multiline"
		],
	}
};
```

.babelrc
```json
{
  "plugins": [
    "transform-flow-comments",
    "transform-function-bind",
    "transform-async-to-generator",
    "transform-exponentiation-operator",
    "transform-class-properties",
    "transform-es2015-destructuring",
    "transform-es2015-spread",
    "transform-es2015-parameters",
    "transform-object-rest-spread",
    "transform-es2015-classes",
    "transform-es2015-modules-commonjs"
  ]
}
```

**What did you do? Please include the actual source code causing the issue.**
I created a "minimal" demo to reproduce the problem. My impression is that since 3.8.0 (it works on 3.7.1) comma-dangle (set to always-multiline) reports a comma in a flow annotation when a object rest parameter is used before.

Complete demo code:
https://gist.github.com/mormahr/e6af53438a9bae9b9fe8905cc938b77d

```js
export default class AntragsEvent {
	id: ?string
	rest: any

	constructor({
		id,
		...rest, // <- this (the object rest spread) is causing
	}: {
		id?: string, // <- this comma to be reported by comma-dangle
	}) {
		this.id = id
		this.rest = rest
	}
}
```

**What did you expect to happen?**
No report of an unexpected comma

**What actually happened? Please include the actual, raw output from ESLint.**

[PRIVATE]/CommaDangleBug/test.js
  9:14  error  Unexpected trailing comma  comma-dangle

✖ 1 problem (1 error, 0 warnings)

**Comment by not-an-aardvark** [comment]
Thanks for the report. This is probably caused by https://github.com/eslint/eslint/commit/c8796e98af6804cccc947ab3e3b06d884a8d035f, or maybe https://github.com/eslint/eslint/commit/52da71e9bfcdb7f13c6077531b992d059af7391a.

**Comment by not-an-aardvark** [comment]
For what it's worth, I think your syntax is invalid based on the [current version of the rest/spread spec](https://sebmarkbage.github.io/ecmascript-rest-spread/); trailing commas aren't allowed after rest properties, since no property could actually be added after it. (See https://github.com/eslint/eslint/issues/7297 for more info.)

```js
class Foo {
  constructor ({
    bar,
    ...baz, // <-- trailing comma is not allowed here according to 
  }) {}
}
```

**Comment by mormahr** [comment]
You're right about the comma after spread, but the comma dangle error persists even after removing it.

**Comment by mormahr** [comment]
Thanks for the fast response and fix 👍🏻

**Comment by sstur** [comment]
Yes, I can confirm this is an issue when upgrading to v3.8.0. Here is a much simpler repro example:
```
let {
  foo,
  ...otherStuff,
} = object;
```

Let me know if you need my eslint config for this.

**Comment by not-an-aardvark** [comment]
@sstur That part is actually intended; your code's syntax is invalid, so it wouldn't make sense for ESLint to require a trailing comma in that case. The error reported there is a result of the bugfix in https://github.com/eslint/eslint/issues/7297.

**Comment by sstur** [comment]
Oh, I see. I believe my syntax is valid for Stage 3 of ECMAScript and is handled for us by Babel. We use ESLint with some newer ECMAScript proposed features such as this (specifically we use the same ES feature set as ReactNative.) It has been working as long as I can remember but broke when upgrading from 3.7.1 to 3.8.0 (presumably because of the bugfix you referenced).

I am sticking to 3.7.1 now but probably need a way forward with proposed, not-yet-standardized features.

**Comment by not-an-aardvark** [comment]
I understand that you're using babel, but a trailing comma after a rest property is not valid syntax even according to the [current proposal spec](https://sebmarkbage.github.io/ecmascript-rest-spread/), because there is nothing that could follow a rest property.

The same thing applies with the already-standardized rest elements:

```js
const [
  foo,
  ...bar, // <-- This is a syntax error in ES6
] = baz;
```

In order to be more compliant with the spec, babel's parser [temporarily reported this as a syntax error](https://github.com/babel/babylon/releases/tag/v6.11.3), but it was [later reverted](https://github.com/babel/babylon/releases/tag/v6.11.4) due to breakage.

The fix in this case is to simply remove the comma, so that it becomes valid syntax according to the Stage 3 proposal.

**Comment by sstur** [comment]
Oh, I see! This I did not know!

So Babel is being more liberal than the standards in what it accepts. Interesting.

Thank you for taking the time to explain this. I wonder how to get the word out there so people don't continue opening issues..
